documentation/kernel-semantics.tex

```diff
@@ -3174,14 +3214,113 @@ \subsubsection{Async}
 
 Along other things, the following is not considered in the basic version:
 \begin{itemize}
-    \item The labels and switch labels are omitted.
+    \item Exception handlers should be updated with a handler that completes the future.
 
-    \item The exception-related components and transitions are omitted.
+    \item Semantics for invocations of async functions other then static invocations should be added.
 
-    \item For simplicity, all functions are considered to have exactly one required positional parameter and zero named parameters.
+    \item Elements from $\textbf{FutureValue}$ should be specified.
 
 \end{itemize}
 
+\renewcommand{\StartNK}[4]{\mathrm{StartNK}({#1},\,{#2},\,{#3},\,\ExceptionHandlersRest,\,{#4})}
+\newcommand{\CompleteFutureEK}[1]{\mathrm{CompleteFutureEK}({#1})}
+\newcommand{\AwaitReturnNK}[2]{\mathrm{AwaitReturnNK}({#1},\,{#2})}
+\newcommand{\AwaitEK}{\mathrm{AwaitEK}(\econt)}
+\newcommand{\AsyncStaticInvA}[4]{\mathrm{AsyncStaticInvA}({#1},\,{#2},\,{#3},\,\ExceptionHandlersRest,\,{#4})}
+
+If await continuation, $\AwaitEK$ is applied to a FutureValue, it needs to "complete" the future before returning, 
+Therefore, it adds an event continuation at the tail of the loop that captures the return continuation and the location from the store corresponding to the returned future.
+It then applies the head event continuation of the event loop.
+
+When the body of async function is about to be executed, we create a $\StartNK{\stmt}{\env}{\strace}{\econt'}$ event continuation that captures the body, the environment, the exception components and an expression continuation that completes the corresponding future.
+We append this event continuation to the event loop and the invocation evaluates to a future immediately.
+The created future value captures the location where the completed value of the future will eventually be stored.
```

Comment:
Just for the reference I'm putting here a note about alternative implementation of `Future`s.  We may treat expression continuations that are awaiting values from `Future`s as some sort of return continuations for the bodies of methods that should produce these awaited values.  When the value is finally produced (returned), we may return it immediately or we may create a new element of the event loop that will return it to the awaiting continuation.  This way, we avoid using an auxiliary location for the returned value.  Additionally, the awaiting continuations don't appear on the event loop until the value is ready, and we don't need to reschedule them on the event loop, thus avoiding the need for the special "not ready" value.